### PR TITLE
feat(TD-8): Add and fetch tasks from backend

### PR DESCRIPTION
**Pull request by taiman-siddiqui-sazim**
**Description of change**

- User can add a task to the backend
- Added task/s get fetched from the backend

**Associated Trello link**

https://trello.com/c/c4vCZIbG

**Screenshot of change**

- Add the following task to the backend: 

Screenshot 2024-12-19 at 6 08 24 AM

- Added task is fetched from the backend:

Screenshot 2024-12-19 at 6 10 45 AM

**Dev notes** 

Features to edit, delete and mark completion of tasks has not been implemented yet.

### CODE REVIEW

File: modules/to-do/components/AddTask/AddTask.tsx [hunk 1]
```diff
@@ -19,13 +20,23 @@ export const AddTask = ({ onSubmit }: TAddTaskProps) => {
...
-    setTimeout(() => setShowPopup(false), 2000);
+      if (newTask?.id) {
+        onTaskAdded(newTask.id);
+        setTitle(""); 
+        setShowPopup(true); 
+      } else {
+        throw new Error("Invalid task data returned from API.");
+      }
+
+      setTimeout(() => setShowPopup(false), 2000);
```
Comment (nur-zaman-sazim) [inline modules/to-do/components/AddTask/AddTask.tsx:35]: ooc: why are we using timeout here ?

File: shared/utils/TaskApi.ts [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+      throw new Error("No response from server. Please try again.");
+    } else {
+      throw new Error("An error occurred while setting up the request.");
+    }
+  }
+};
+
+export const fetchTasksFromApi = async (taskId?: number): Promise<ITask | ITask[]> => {
+    try {
+      const url = taskId ? `${API_BASE_URL}/${taskId}` : API_BASE_URL; // Conditional endpoint
```
Comment (mahir-mohsin-sazim) [inline shared/utils/TaskApi.ts:27]: Remove comments

File: shared/typedefs/constants.ts [hunk 3]
```diff
@@ -1 +1,2 @@
 export const MAX_TITLE_LENGTH = 40;
+export const API_BASE_URL = "http://localhost:3001/api/v1/tasks";
```
Comment (mahir-mohsin-sazim) [inline shared/typedefs/constants.ts:2]: Shouldn't this be in `.env` file??

File: modules/to-do/containers/ToDoPageContainer.tsx [hunk 4]
```diff
@@ -1,48 +1,43 @@
...
-  useEffect(() => {
-    const storedTasks = localStorage.getItem("tasks");
-    if (storedTasks) {
-      setTasks(JSON.parse(storedTasks));
+  const handleTaskAdded = async (taskId: number) => {
+    try {
+      const newTask = await fetchTasksFromApi(taskId);
+      setSingleTask(newTask as ITask); 
+    } catch (error) {
+      console.error("Error fetching added task:", error);
```
Comment (mahir-mohsin-sazim) [inline modules/to-do/containers/ToDoPageContainer.tsx:16]: Just a quick reminder, for the next project you need to do proper error handling. `console.loging` the error won't be acceptable.

**Review by mahir-mohsin-sazim — CHANGES_REQUESTED** [review]

File: modules/to-do/components/AddTask/AddTask.tsx (hunk 1 above)
Comment (taiman-siddiqui-sazim) [reply modules/to-do/components/AddTask/AddTask.tsx:35]: To display the popup after adding task for 2 seconds.

**Pull request merged**